From ed0f3be370d9d94d62f9da2442d407ab1cdb988c Mon Sep 17 00:00:00 2001
From: Liu Ying <Ying.Liu@freescale.com>
Date: Wed, 21 May 2014 13:36:02 +0800
Subject: [PATCH 0828/1074] ENGR00313867 media: mxc: ADV7180: Get standard
 when the chip is locked

The IN_LOCK bit of the ADV7180 register STATUS_1(0x10) indicates if
the chip is locked or not in auto detection mode.  We should get
valid standard when the chip is locked, otherwise, invalid standard
should be returned.  This patch checks the IN_LOCK bit when we get
standard in auto detection mode.

Conflicts:

	drivers/media/video/mxc/capture/adv7180.c

Signed-off-by: Liu Ying <Ying.Liu@freescale.com>
(cherry picked from commit c0d94e51b331002c11ebdfc8a559c8ec878894d6)
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 drivers/media/platform/mxc/capture/adv7180.c |   34 ++++++++++++-------------
 1 files changed, 16 insertions(+), 18 deletions(-)

diff --git a/drivers/media/platform/mxc/capture/adv7180.c b/drivers/media/platform/mxc/capture/adv7180.c
index ae17d7f..3e3b545 100644
--- a/drivers/media/platform/mxc/capture/adv7180.c
+++ b/drivers/media/platform/mxc/capture/adv7180.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2005-2013 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2005-2014 Freescale Semiconductor, Inc. All Rights Reserved.
  */
 
 /*
@@ -311,28 +311,26 @@ static int adv7180_write_reg(u8 reg, u8 val)
  */
 static void adv7180_get_std(v4l2_std_id *std)
 {
-	int tmp;
-	int idx;
+	int status_1, standard, idx;
+	bool locked;
 
 	dev_dbg(&adv7180_data.sen.i2c_client->dev, "In adv7180_get_std\n");
 
-	/* Read the AD_RESULT to get the detect output video standard */
-	tmp = adv7180_read(ADV7180_STATUS_1) & 0x70;
+	status_1 = adv7180_read(ADV7180_STATUS_1);
+	locked = status_1 & 0x1;
+	standard = status_1 & 0x70;
 
 	mutex_lock(&mutex);
-	if (tmp == 0x40) {
-		/* PAL */
-		*std = V4L2_STD_PAL;
-		idx = ADV7180_PAL;
-	} else if (tmp == 0) {
-		/*NTSC*/
-		*std = V4L2_STD_NTSC;
-		idx = ADV7180_NTSC;
-	} else {
-		*std = V4L2_STD_ALL;
-		idx = ADV7180_NOT_LOCKED;
-		dev_dbg(&adv7180_data.sen.i2c_client->dev,
-			"Got invalid video standard!\n");
+	*std = V4L2_STD_ALL;
+	idx = ADV7180_NOT_LOCKED;
+	if (locked) {
+		if (standard == 0x40) {
+			*std = V4L2_STD_PAL;
+			idx = ADV7180_PAL;
+		} else if (standard == 0) {
+			*std = V4L2_STD_NTSC;
+			idx = ADV7180_NTSC;
+		}
 	}
 	mutex_unlock(&mutex);
 
-- 
1.7.5.4

